File: src/logic/word_pool.py
import random


class word_pool:

    # list containing the word_pool
    _wp = []
    # size of the word_pool
    _sz = 0
    # String containing the path to the file containing the words to load
    _fp = ""


    '''
        Constructor

        Parameter:  None
        Returns:    bool
    '''
    def __init__(self, file_path):
        self._fp = file_path
        self.load_word_pool()




    '''
    Method used to load the word_pool

        Parameter:  None
        Returns:    None
    '''
    def load_word_pool(self):
        print("word_pool reloaded")
        self._wp = open(self._fp).read().splitlines()
        self._sz = len(self._wp)




    '''
    Method used return a random word from the word_pool

        Parameter:  None
        Returns:    str
    '''
    def get_word(self) -> str:
        word = ""

        # if the word_pool is exhausted, it is reloaded
        if self.is_empty():
            self.load_word_pool()
            self._sz = len(self._wp)

        # getting a random word from the word pool
        # removing it from the list
        rng         = random.randrange(0, self._sz)
        word        = self._wp[rng]
        self._sz   -= 1
        del self._wp[rng]

        return word




    '''
        Method used to know if the word_pool has been exhausted

        Parameter:  None
        Returns:    bool
    '''
    def is_empty(self) -> bool:
        return self._sz == 0




    '''
        Method used to test the word pool reloading

        Parameter:  None
        Returns:    None
    '''
    def test_get_word(self):
        for i in range(0, 10):
            print(self.get_word())

File: rust/src/word_pool.rs


extern crate rand;
use rand::Rng;



pub struct WordPool {
    word_pool: Vec<String>,
    file_path: String,
    size:      usize
}


impl WordPool {


    /// Constructor of the WordPool structure
    ///
    /// It requires a file_path to a file containing the words the player will have to find.
    /// 
    /// ```
    ///    let wp: WordPool = WordPool::new(FILEPATH);
    ///
    ///    assert_eq!(wp.word_pool.contains(&"ZONES".to_string()), true);
    ///    assert_eq!(wp.word_pool.contains(&"ZOO-LOGIE".to_string()), true);
    ///    assert_eq!(wp.word_pool.contains(&"ZOZOTER".to_string()), true);
    ///    assert_eq!(wp.word_pool.contains(&"ZYGOTE".to_string()), true);
    ///
    /// ```
    pub fn new(filepath: String) -> WordPool{
        WordPool{
            file_path: filepath,
            word_pool: Vec::new(),
            size: 0
        }
    }


    /// Opens and reads the file containing the words into a vector of strings
    pub fn load_word_pool(self: &mut Self) {
        use std::fs;

        let err_msg = format!("file {} wasn't found", self.file_path);
        let contents = fs::read_to_string(self.file_path.clone())
            .expect(&err_msg);

        self.word_pool = contents.split_whitespace().map(str::to_string).collect();

        self.size = self.word_pool.len();
    }



    /// This method returns a random word from the word_pool,
    /// that the player will have to find.
    /// 
    /// This word will be taken out of the word_pool so that it won't be
    /// picked a second time.
    /// 
    /// # Example
    /// ```
    ///    const WORD_LIST: [&str; 4] = ["ZONES", "ZOO-LOGIE", "ZOZOTER", "ZYGOTE"];
    ///
    ///    let mut wp: WordPool = WordPool::new(FILEPATH.to_string());
    ///
    ///    for _ in 0..wp.size {
    ///        assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
    ///    }
    ///
    ///    // testing one last time to make sure the file is reloaded
    ///    assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
    /// ```
    pub fn get_word(self: &mut Self) -> String {
        // if the word_pool is empty, then we reload it
        if self.is_empty() {
            self.load_word_pool();
        }

        let mut rng = rand::thread_rng();
        let n: u32 = rng.gen_range(0..self.size as u32);
        println!("{}", n);

        self.size -= 1;

        self.word_pool.swap_remove(n as usize)
    }


    /// This method returns true if the word_pool is empty, false otherwise
    ///
    /// # Example
    /// ```
    ///    let mut wp: WordPool = WordPool::new(FILEPATH);
    ///
    ///    let wl = WORD_LIST.len(); 
    ///    let wp_size = wp.word_pool.len();
    ///    for i in 0..wp_size {
    ///        let size = wp.word_pool.len();
    ///        assert_eq!(size, wl - i);
    ///        assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
    ///    }
    ///    // testing one last time to make sure the file is reloaded
    ///    assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
    /// ```
    pub fn is_empty(self: &Self) -> bool {
        self.size == 0
    }


}


#[cfg(test)]
mod tests {
    use super::WordPool;

    const FILEPATH: &str = "../files/test";

    #[test]
    fn test_new() {
        let wp: WordPool = WordPool::new(FILEPATH.to_string());

        assert_eq!(wp.word_pool.len(), 0);
    }


    #[test]
    fn test_is_empty() {
        let mut wp: WordPool = WordPool::new(FILEPATH.to_string());

        assert_eq!(wp.is_empty(), true);
        wp.load_word_pool();
        assert_eq!(wp.is_empty(), false);

    }


    #[test]
    fn test_get_word() {
        const WORD_LIST: [&str; 4] = ["ZONES", "ZOO-LOGIE", "ZOZOTER", "ZYGOTE"];

        let mut wp: WordPool = WordPool::new(FILEPATH.to_string());

        for _ in 0..wp.size {
            assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
        }

        // testing one last time to make sure the file is reloaded
        assert_eq!(WORD_LIST.contains(&wp.get_word().as_str()), true);
    }

}